url0067cdd33d_goldolphin_Mi_tgz-pJ8-mi_regex_CommonJ8/src/mi/match/ACSearcher.java

```diff
@@ -0,0 +1,138 @@
+package mi.match;
+
+import mi.common.CharHashMap;
+import mi.common.Utils;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+
+/**
+ * User: goldolphin
+ * Time: 2013-06-12 17:25
+ */
+public class ACSearcher {
+    private Node root;
+
+    public ACSearcher(Collection<String> patterns) {
+        root = new Node();
+        for (String pattern: patterns) {
+            add(root, pattern, 0);
+        }
+        buildFails(root, root);
+    }
+
+    public void search(String source, IMatchHandler handler) {
+        int sourceLen = source.length();
+        int pos = 0;
+        Node current = root;
+        while (true) {
+            if (current.pattern != null) {
+                handler.handle(pos, current.pattern);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;handler&#39; (index: 1 | type: IMatchHandler). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other validated parameters: &#39;source&#39;. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).